crates/extensions/c8y_mapper_ext/src/availability.rs

```diff
@@ -0,0 +1,255 @@
+use crate::actor::SyncStart;
+use crate::actor::TimeoutKind;
+use crate::service_monitor::get_health_status_from_message;
+use crate::service_monitor::HealthStatus;
+use std::collections::HashMap;
+use std::time::Duration;
+use tedge_actors::ChannelError;
+use tedge_actors::LoggingSender;
+use tedge_actors::Sender;
+use tedge_api::entity_store::EntityMetadata;
+use tedge_api::mqtt_topics::Channel;
+use tedge_api::mqtt_topics::EntityTopicId;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::mqtt_topics::ServiceTopicId;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::Topic;
+use tracing::debug;
+use tracing::info;
+
+pub type TopicWithoutPrefix = String;
+
+/// The timer payload for TimeoutKind::Heartbeat.
+///
+/// `device` should hold the EntityTopicId of the device for availability monitoring
+/// `health` should hold the Topic of the device's lead service.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct HeartbeatPayload {
+    pub device: EntityTopicId,
+    pub health: TopicWithoutPrefix,
+}
+
+pub fn create_c8y_heartbeat_message(
+    map: &HashMap<TopicWithoutPrefix, HealthStatus>,
+    c8y_topic: &Topic,
+    heartbeat: &HeartbeatPayload,
+) -> Option<MqttMessage> {
+    let health_topic_name = heartbeat.health.clone();
+
+    match map.get(&health_topic_name) {
+        Some(health_status) if health_status.status == "up" => {
+            Some(MqttMessage::new(c8y_topic, "{}"))
+        }
+        _ => {
+            debug!("Heartbeat message is not generated because the status of '{health_topic_name}' is not 'up'");
+            None
+        }
+    }
+}
+
+pub async fn set_heartbeat_timer_from_metadata(
+    period: i16,
+    status_map: &mut HashMap<TopicWithoutPrefix, HealthStatus>,
+    timer_map: &mut HashMap<EntityTopicId, TopicWithoutPrefix>,
```

Comment:
Exactly the intention is not to expand `converter.rs`. Fed up with scrolling the long lines and addressing conflicts when someone wants to refactor the file.

> If we had something like an AvailabilityManager struct with ownership of those maps in this module, operating on them independently, it would have been better.

We can discuss it later.